### PR TITLE
Pymysql cursor types and fetching return types

### LINKED ISSUE #5567: pymysql: fetch result type
```python
import pymysql

connection = pymysql.connect(password='hunter2', cursorclass=pymysql.cursors.DictCursor)
cursor = connection.cursor()
reveal_type(cursor)  # pymysql.cursors.Cursor, but DictCursor at runtime
reveal_type(cursor.fetchone())  # Union[builtins.tuple[Any], builtins.dict[builtins.str, Any], None]

cursor.execute("select 1 as foo")
foo = cursor.fetchone()['foo']  # error: Value of type "Union[Tuple[Any, ...], Dict[str, Any], None]" is not indexable
```

Ideally `cursor` would have type `pymysql.cursors.DictCursor`, so that `cursor.fetchone()` would return `Dict[str, Any]` instead of an annoying union.

Because `cursor.fetchone()` can return `None` but often checking for it is annoying, it could instead return `Union[Dict[str, Any], Any]`. See #5528 for a similar situation.

We could make `connection` generic and overload `connect()` to figure out what type `connection.cursor()` returns. Something like this (omitting lots of irrelevant stuff):

```python
_CursorT = TypeVar("_CursorT", bound=pymysql.cursors.Cursor)

class Connection(Generic[_CursorT]):
   def cursor(self) -> _CursorT: ...

@overload
def connect(*, cursorclass: Type[_CursorT]) -> Connection[_CursorT]: ...
@overload
def connect(*, cursorclass: Type[pymysql.cursors.Cursor] = ...) -> Connection[pymysql.cursors.Cursor]: ...
```

The downside is that now all other arguments of `connect()` have to be copy/pasted into two places, and there's many of them.

### PR DESCRIPTION
**Pull request by Akuli**
Fixes #5567 so that `Cursor` and `DictCursor` no longer return annoying unions. I also got rid of several `Optional`s. In those methods, I looked at the source code and I tried `select * from foo` on a local test database, and couldn't get it to return `None`.

I also made `Connection` generic, so that when you create a connection, it remembers which cursor class you want to use. You can still customize this by passing a custom class to `.cursor()`.

### CODE REVIEW

File: stubs/PyMySQL/pymysql/connections.pyi [hunk 1]
```diff
@@ -71,8 +74,53 @@ class Connection:
     init_command: Any
     max_allowed_packet: int
     server_public_key: bytes
+    @overload
+    def __init__(
+        self: Connection[Cursor],  # different between overloads
+        *,
+        host: Optional[str] = ...,
```
Comment (srittau) [inline stubs/PyMySQL/pymysql/connections.pyi:81]: `str | None` (also below)

File: stubs/PyMySQL/pymysql/connections.pyi [hunk 2]
```diff
@@ -71,8 +74,53 @@ class Connection:
...
+        cursorclass: None = ...,  # different between overloads
+        init_command: Optional[Any] = ...,
+        connect_timeout: Optional[int] = ...,
+        ssl: Mapping[Any, Any] | None = ...,
+        ssl_ca: Any = ...,
+        ssl_cert: Any = ...,
+        ssl_disabled: Any = ...,
+        ssl_key: Any = ...,
+        ssl_verify_cert: Any = ...,
+        ssl_verify_identity: Any = ...,
```
Comment (srittau) [inline stubs/PyMySQL/pymysql/connections.pyi:102]: Don't use `Any` here, just leave unannotated, same in the other overload.

**Review by srittau — CHANGES_REQUESTED** [review]
I haven't compared with the implementation, but the changes look reasonable. Especially getting rid of the response unions. Two remarks below.

Also, could you also update the version in `METADATA.toml`?

**Review by srittau — APPROVED** [review]

**Comment by Alch-Emi** [comment]
I'm having some problems with this, but I'm posting here instead of creating a new issue in case I'm doing something dumb.  I'm currently writing some typed code that uses a `Connection[Cursor]`, but when I include this in my function signature, I get the error `TypeError: 'type' object is not subscriptable`, presumably since the underlying library doesn't expect `Connection` to be a generic.

I was able to fix this either by patching pymysql to use a generic Connection, or by removing the generic from my type signature.  The former solution of course makes the code much harder to maintain, and would require a fork of pymysql specifically for this project, at least until pymysql adopts python 3.10 or 3.7 (See https://github.com/PyMySQL/PyMySQL/issues/953).  The later causes my typechecker to complain very loudly about how Connection is supposed to be a generic, and also removes the benefit of having the generic in the first place.

Is there a solution to this that doesn't require either leaving a type error in the code or forking pymysql?

**Comment by Akuli** [comment]
On 3.7+, add to beginning of file:

```python
from __future__ import annotations
```

This tells Python to not evaluate the type annotations, so that they are just for the type checker. It also means that you can also use `Foo | None` instead of `Optional[Foo]`, for example.

**Comment by Alch-Emi** [comment]
Oh thank you so much!  That's perfect!

**Pull request merged**